Core/src/Geometry/DetectorVolume.cpp

```diff
@@ -0,0 +1,228 @@
+  if (pIndex >= m_portals.internal.size()) {
+    throw std::invalid_argument(
+        "DetectorVolume: trying to update a portal that does not exist.");
+  }
+  m_portals.internal[pIndex] = portal;
+  m_portals = ObjectStore<std::shared_ptr<Portal>>(m_portals.internal);
+}
+
+void Acts::Experimental::DetectorVolume::construct(
+    const GeometryContext& gctx, const PortalGenerator& portalGenerator) {
+  // Create portals with the given generator
+  auto portalSurfaces =
+      portalGenerator(transform(gctx), *(m_bounds.get()), getSharedPtr());
+  m_portals = ObjectStore<std::shared_ptr<Portal>>(portalSurfaces);
+}
+
+std::shared_ptr<Acts::Experimental::DetectorVolume>
+Acts::Experimental::DetectorVolume::getSharedPtr() {
+  return shared_from_this();
+}
+
+std::shared_ptr<const Acts::Experimental::DetectorVolume>
+Acts::Experimental::DetectorVolume::getSharedPtr() const {
+  return shared_from_this();
+}
+
+bool Acts::Experimental::DetectorVolume::inside(const GeometryContext& gctx,
+                                                const Vector3& position,
+                                                bool excludeInserts) const {
+  Vector3 posInVolFrame((transform(gctx).inverse()) * position);
+  if (not volumeBounds().inside(posInVolFrame)) {
+    return false;
+  }
+  if (not excludeInserts or m_volumes.external.empty()) {
+    return true;
+  }
+  // Check exclusion through subvolume
+  for (const auto v : volumes()) {
+    if (v->inside(gctx, position)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void Acts::Experimental::DetectorVolume::updateNavigationState(
+    const GeometryContext& gctx, NavigationState& nState) const {
+  // This volume still has sub volumes, it needs a double-check
+  // if the internal volumes can be excluded.
+  //
+  // This can be avoided by a precise volume finder
+  if (m_volumes.external.size()) {
+    for (const auto v : volumes()) {
+      if (v->inside(gctx, nState.position)) {
+        v->updateNavigationState(gctx, nState);
+        return;
+      }
+    }
+  }
+  // The state updator of the volume itself
+  nState.currentVolume = this;
+  m_SurfaceCandidatesUpdator.delegate(gctx, nState);
+  nState.surfaceCandidate = nState.surfaceCandidates.begin();
+  return;
+}
+
+void Acts::Experimental::DetectorVolume::assignSurfaceCandidatesUpdator(
+    ManagedSurfaceCandidatesUpdator&& navStateUpdator,
+    const std::vector<std::shared_ptr<Surface>>& surfaces,
+    const std::vector<std::shared_ptr<DetectorVolume>>& volumes) {
+  m_SurfaceCandidatesUpdator = std::move(navStateUpdator);
+  m_surfaces = ObjectStore<std::shared_ptr<Surface>>(surfaces);
+  m_volumes = ObjectStore<std::shared_ptr<DetectorVolume>>(volumes);
+
+  // This can take a while for complicated geometries
+  // it also needs a geometry context, needs to be figured out
+  // if (not checkContainment(GeometryContext)) {
+  //  throw std::runtime_error("Objects are not contained by volume.");
+  //}
```

Comment:
dead?

Reply:
Not dead, but a fundamental problem:

I wanted to do that in an `assert()`, but for the containment I need the `GeometryContext`, so if I give a context object name in the constructor I get an `unused variable` WARNING, if I add [[maybe_unused]] in the header, the CUDA build fails ...

Reply:
Have you tried a workaround like this? Maybe the CUDA compiler has a problem with the attribute in the function argument list?

```c++
DetectorVolume::DetectorVolume(const GeometryContext &gctx, ...) {
    // ...
    [[maybe_unused]] const auto &gctx_ref = gctx;
    assert(checkContainment(gctx) && "Objects are not contained by volume.");
    // ...
}
```

Reply:
I will try this, that'd be a good workaround.